Code/GraphMol/Chirality.cpp

```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
-                atom->getChiralTag()) {
-              same = -1;
+    if (!nextAtoms.empty()) {
+      atom->getPropIfPresent(common_properties::_ringStereoAtoms,
+                             ringStereoAtoms);
+    }
+
+    while (!nextAtoms.empty()) {
+      const Atom *ratom = nextAtoms.front();
+      nextAtoms.pop_front();
+      atomsSeen.set(ratom->getIdx());
+      if (ratom->getChiralTag() != Atom::CHI_UNSPECIFIED &&
+          !ratom->hasProp(common_properties::_CIPCode) &&
+          atomIsCandidateForRingStereochem(mol, ratom)) {
+        int same = (ratom->getChiralTag() == atom->getChiralTag()) ? 1 : -1;
+        ringStereoAtoms.push_back(same * (ratom->getIdx() + 1));
+        INT_VECT oringatoms(0);
+        ratom->getPropIfPresent(common_properties::_ringStereoAtoms,
+                                oringatoms);
+        oringatoms.push_back(same * (atom->getIdx() + 1));
+        ratom->setProp(common_properties::_ringStereoAtoms, oringatoms, true);
+        possibleSpecialCases.set(ratom->getIdx());
+        possibleSpecialCases.set(atom->getIdx());
+      }
+      // now push this atom's neighbors
+      boost::tie(beg, end) = mol.getAtomBonds(ratom);
+      while (beg != end) {
+        unsigned int bidx = mol[*beg]->getIdx();
+        if (!bondsSeen[bidx]) {
+          bondsSeen.set(bidx);
+          if (mol.getRingInfo()->numBondRings(bidx)) {
+            const Atom *oatom = mol[*beg]->getOtherAtom(ratom);
+            if (!atomsSeen[oatom->getIdx()] && !atomsUsed[oatom->getIdx()]) {
+              nextAtoms.push_back(oatom);
+              atomsUsed.set(oatom->getIdx());
             }
-            ringStereoAtoms.push_back(same * (*idxIt + 1));
-            INT_VECT oAtoms(0);
-            mol.getAtomWithIdx(*idxIt)->getPropIfPresent(
-                common_properties::_ringStereoAtoms, oAtoms);
-
-            oAtoms.push_back(same * (atom->getIdx() + 1));
-            mol.getAtomWithIdx(*idxIt)->setProp(
-                common_properties::_ringStereoAtoms, oAtoms, true);
           }
         }
+        ++beg;
       }
-    }
-    if (ringStereoAtoms.size()) {
+    }  // end of BFS
+    if (ringStereoAtoms.size() != 0) {
       atom->setProp(common_properties::_ringStereoAtoms, ringStereoAtoms, true);
-      return true;
+      // because we're only going to hit each ring atom once, the first atom we
+      // encounter in a ring is going to end up with all the other atoms set as
+      // stereoAtoms, but each of them will only have the first atom present. We
+      // need to fix that. because the traverse from the first atom only
+      // followed ring bonds, these things are all by definition in one ring
+      // system. (Q: is this true if there's a spiro center in there?)
+      INT_VECT same(mol.getNumAtoms(), 0);
+      BOOST_FOREACH (int ringAtomEntry, ringStereoAtoms) {
+        int ringAtomIdx =
+            ringAtomEntry < 0 ? -ringAtomEntry - 1 : ringAtomEntry - 1;
```

Comment:
When tagging atoms for special purposes, I tend to write inline functions to help me cognitively.  This is a style thing so I don't expect agreement.  Here is an example, however:
```
inline int tagIdxWithSameChirality(unsigned int atomIdx, bool sameChirality) {
  return ((int)sameChirality * rdcast<int>(atomIdx + 1));
}

inline int getIdxFromTax(int tag) {
  return tag > 0 ? tag - 1 : -(tag - 1) 
}

inline bool hasSameChirality(int tag) {
   return tag > 0;
}
```

And here is the confusing usage:

```
BOOST_FOREACH (int ringAtomEntry, ringStereoAtoms) {
        same[getIdxFromTag[ringAtomEntry]] = hasSameChirality(ringAtomEntry);
      }
```

Which I find reads a lot better and should have the same performance.  Just my 2cents though.  Now that I understand the code, It seems to solve the current issue.